### PR TITLE
bh/jh - Add image tags

### LINKED ISSUE #6: Add Style
#### As a user I want to see a visually pleasing application.

Acceptance criteria:
- [x] The application has a header with navigation
- [ ] Images are displayed in an aesthetically pleasing list of cards
- [ ] Form errors are shown inline with the related input field
- [ ] I know how to use Bootstrap components like buttons and navbars in my
  application.
- [ ] I know the very basics of CSS and SCSS

Reference material:
- http://v4-alpha.getbootstrap.com/getting-started/introduction/
- https://github.com/plataformatec/simple_form

Dependencies:
- Image Index

### PR DESCRIPTION
**Pull request by blakeherrington**
Closes #6

### CODE REVIEW

File: db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
```
Comment (shfung) [inline db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Maybe we can add some not-null constraints to some of the columns?

File: test/flow_tests/images_flow_test.rb [hunk 2]
```diff
@@ -3,20 +3,24 @@
...
     assert_selector('#new_image', count: 1)
 
     fill_in('Url', with: 'invalid url')
+    fill_in('Tag list', with: 'blah')
     click_button('Create Image')
     assert_selector('#new_image', count: 1)
     assert_selector('.error', text: 'must be a valid URL')
 
     fill_in('Url', with: valid_url)
+    fill_in('Tag list', with: test_tag_list)
```
Comment (shfung) [inline test/flow_tests/images_flow_test.rb]: Maybe you can add another test for the scenario where no tag is given?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -17,6 +18,7 @@ class ImagesControllerTest < ActionController::TestCase
 
     image = Image.last
     assert_equal test_image_url, image.url
+    assert_equal test_tag_list, image.tag_list.to_s
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Similar to my previous comment, maybe you can add another test for the scenario where no tag is given?

File: db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (blakeherrington) [inline db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: These are actually the migrations generated by acts-on-taggable when I installed the gem, I suppose I could still change it though, if that's what you suggest? :)

Comment (shfung) [inline db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Yes, it's generally a good practice to add not-null constraints to all columns except the ones that actually allow null.

Comment (blakeherrington) [inline db/migrate/20160407215312_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Great! I'll go ahead and do that with my other changes. Committing shortly.

File: test/flow_tests/images_flow_test.rb (hunk 2 above)
Comment (blakeherrington) [inline test/flow_tests/images_flow_test.rb]: Done

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Makes sense - also done

**Comment by blakeherrington** [comment]
I think I addressed the things you mentioned - as well as fixing some pronto conflicts. Let me know when you get a chance to take a second look. Thank you!

File: app/views/images/index.html.erb [hunk 4]
```diff
@@ -6,4 +6,7 @@
 
 <% @images.each do |image| %>
   <%= image_tag image.url %>
+  <div class='tag-list'>
```
Comment (rmacklin) [inline app/views/images/index.html.erb]: Please use double quotes around HTML attributes.

Comment (blakeherrington) [inline app/views/images/index.html.erb]: Done!

**Comment by blakeherrington** [comment]
@shfung Any other comments on this branch before I merge? Thanks!

File: db/migrate/20160407215314_add_taggings_counter_cache_to_tags.acts_as_taggable_on_engine.rb [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # This migration comes from acts_as_taggable_on_engine (originally 3)
 class AddTaggingsCounterCacheToTags < ActiveRecord::Migration
   def self.up
-    add_column :tags, :taggings_count, :integer, default: 0
+    add_column :tags, :taggings_count, :integer, default: 0, null: false
```
Comment (shfung) [inline db/migrate/20160407215314_add_taggings_counter_cache_to_tags.acts_as_taggable_on_engine.rb]: It is usually better to keep all changes of a single migration file within a same commit.  Can you move these migration changes to the previous commit?

**Comment by shfung** [comment]
Looks good. Just one more comment about the commits. Thanks.

File: db/migrate/20160407215314_add_taggings_counter_cache_to_tags.acts_as_taggable_on_engine.rb (hunk 5 above)
Comment (blakeherrington) [inline db/migrate/20160407215314_add_taggings_counter_cache_to_tags.acts_as_taggable_on_engine.rb]: Ah yes, that makes sense. Thanks!

**Comment by rmacklin** [comment]
@shfung:
> Looks good.

I'm still seeing a lot of columns [here](https://github.com/AppFolioOnboarding/image-sharer-blakeherrington/blob/dfa0f1627ae114708fbd955cfc1791302c457196/db/schema.rb) that don't have not-null constraints.

@blakeherrington Where did you add the not-null constraints after [@shfung's comment](https://github.com/AppFolioOnboarding/image-sharer-blakeherrington/pull/24#discussion_r58963927)? Was it just to `taggings_count`?

**Comment by blakeherrington** [comment]
@rmacklin You're right. I didn't complete understand the comment, but now I have changed the migration to have null: false everywhere and schema.rb got updated accordingly. For some reason I originally though it was only necessary on the add_column call :).

**Comment by blakeherrington** [comment]
Oh interesting... getting some errors when trying to apply those migrations... they seemed to work locally just fine. Looking at it now.

**Comment by blakeherrington** [comment]
Got the build fixed. Will be merging to master after I resolve conflicts 👍  Thanks to both of you for all your help!

**Pull request merged**